### DISCUSSION TITLE
How can I get the array key from each entry?

### DISCUSSION THREAD

**Discussion by Truncated** [farling42/obsidian-import-json#72] [Q&A] [answered]
I am trying to use this to import a backup JSON from the Bullet Journal app, which I have found to be helpful in UX for gathering pictures of journal pages and referencing on my mobile. Ideally, I'd like to make a workflow with this that is as few import steps as necessary.

The schema, roughly, is such:
```
{
    "libraryVolumes": [
              { "collections": [
            ],
            "coverPage": {
                "created": "2024-04-13T19:19:02Z",
                "id": "aac574cc-e249-4f62-9595-ec7412e2feb9",
                "image": {
                    "id": "00438789-0324-45dc-ac1a-88cfcd0b8c4b"
                }
            },
            "editionTwoFutureLogPages": [ **<<Sometimes the same kind of entries as "pages" is here.>>** ],
            "editionTwoIndexPages": [ **<<Sometimes the same kind of entries as "pages" is here.>>** ],
            "endDate": "2020-02",
            "firstPageSide": "left",
            "id": "98038d50-c393-4303-abed-fb43e4730173",
            "indexPages": [],
            "pageCount": 240,
            "pages": [
                {
                    "collections": [
                        "Transitions"
                    ],
                    "created": "2024-04-13T19:19:43Z",
                    "date": "2020-01-01",
                    "id": "437f5e3d-b7a8-4c73-b000-534980de3bc0",
                    "page": 1,
                    "image": {
                        "id": "f9da5d23-af08-40b9-9ba8-63596bc31155"
                    }
                },
                {
                    "collections": [
                        "Transitions"
                    ],
                    "created": "2024-04-13T19:19:57Z",
                    "date": "2020-01-01",
                    "id": "b69ea833-b91c-42e6-957c-d459a505e384",
                    "page": 2,
                    "image": {
                        "id": "fff5d72b-7f2e-4d9b-bfc0-79cf6ad74df9"
                    }
                }}
            ],
            "reflections": [],
            "startDate": "2020-01",
            "volumeType": "other"
        },
     {  **<<<Another Journal layout, as above>>>**
     }
    ]
}
```

In pure Handlebars using a VSCode preview, I am using this to parse it getting the elements I want as separate notes with this using #each:
```
{{#each libraryVolumes}}
    {{#gt @key 8}}
    {{#gt @key 98}}
    JUID: J{{add @key 1}}<br>
    {{else}}
    JUID: J0{{add @key 1}}<br>
    {{/gt}}
    {{else}}
    JUID: J00{{add @key 1}}<br>
    {{/gt}}
    Start Date: {{startDate}}<br>
    End Date: {{endDate}}<br>
    First Page Side: {{firstPageSide}}<br>
    {{#is firstPageSide "left"}}
    First Page Left: "true"<br>
    {{else}}
    First Page Left: "false"<br>
    {{/is}}
    Pages: {{pageCount}}<br>
    created: {{coverPage.created}}<br>
    Cover: '__attachments/{{coverPage.image.id}}.jpg'<br>
        {{#each editionTwoIndexPages}}
        tags:<br>
         - 'Indicies'<br>
        Page: {{page}}<br>
        Date: {{@importDataRoot.startDate}}<br>
        Image: '__attachments/{{image.id}}.jpg'<br>
        {{/each}}
        {{#each editionTwoFutureLogPages}}
        tags:<br>
         - 'Future Logs'<br>
        Page: {{page}}<br>
        Date: {{@importDataRoot.startDate}}<br>
        Image: '__attachments/{{image.id}}.jpg'<br>
        {{/each}}
        {{#each pages}}
        tags:<br>
         - '{{collections}}'<br>
        Page: {{page}}<br>
        Date: {{date}}<br>
        Image: '__attachments/{{image.id}}.jpg'<br>
        {{/each}}
        <br>
{{/each}}
```

My understanding is that the `#each` commands can't really be used with this add-on (please correct if I'm wrong on any of my assumptions here or below), so I likely have to import each array individually. Currently I think this means I have to make at least four import passes with the following "Field containing the data" and a separate template file for each specific nuance of each of these four:
libraryVolumes = A Journal
editionTwoIndexPages = A Page
editionTwoFutureLogPages = A Page
pages  = A Page

My template for the journal is:
```
---
fileClass: Journal
Journal Type: "[[120g Paper Edition]]"
{{#gt @importDataRoot 8}}
{{#gt @importDataRoot 98}}
BuJo Name: J{{add @importDataRoot 1}}
{{else}}
BuJo Name: J0{{add @importDataRoot 1}}
{{/gt}}
{{else}}
BuJo Name: J00{{add @importDataRoot 1}}
{{/gt}}
juid: {{id}}
Start Date: {{startDate}}
End Date: {{endDate}}
{{#is firstPageSide "left"}}
First Page Left: true
{{else}}
First Page Left: false
{{/is}}
Pages: {{pageCount}}
created: {{coverPage.created}}
Cover: "[[{{coverPage.image.id}}.jpg]]"
---
![[{{coverPage.image.id}}.jpg]]
```
Importing this with **Data contains multiple JSON objects = true**, **Field containing the data = libraryVolumes**, **Each subfield is a separate note = true**, and **Field to use as Note name = id** does make an entry that I might want.

However @importDataRoot always presents as 0 and "@key" appears to be null. How might I reference the spot it was in the array?
I'd prefer the juid construct to be used as the Note Name as well, but I don't know how that would work even if the juid @key/@importDataRoot worked.

**Comment by Truncated** [comment]
I think ideally I'd like my juid construct to be used in Obsidian, but maintaining the id is critical for updating deltas later on. I've gotten the id to work with the pages, though, using this template:
```
---
fileClass: Page
juid: "[[{{@importDataRoot.id}}]]"
tags:
    - 'Indicies'
Page: {{page}}
Date: {{@importDataRoot.startDate}}
Image: "[[{{image.id}}.jpg]]"
---
![[{{image.id}}.jpg]]
```
The juid here is obviously taking the root, which is the Journal id. The benefit of using the array ID as above is that the constructed juid matches how the BuJu app titles each journal in it's built-in Library. I'd prefer to have them mirrored and not have to upkeep that manually. I anticipate running this on the regular as new journals and pages are added, which I average a rate of 5-10 pages a day with diagrams and analysis for my work. 

Luckily that really only applies to the pages job; the other four are usually one-time for each journal setup.

**Reply by farling42** [reply]
`@importSourceIndex` should contain the current index into the array you selected as the field to iterate over.

`@importDataRoot.libraryVolumes` would be how to access the array which you have chosen to process. `@importDataRoot` is normally used to access something in the selected JSON file that is outside of the array element currently being processed.

**Reply by Truncated** [reply]
> @importSourceIndex should contain the current index into the array you selected as the field to iterate over.

When I replace @importDataRoot in my Journal template with that, it returns J001 for every note made, which indicates to me that either it is returning 0 because it's seeing it as the first entry of an array for every separate note, or it's a null number that gets read as a 0. I added it to the "body" of the page in the template as well and it returns 0 every time.

Is there a setting mis-match that would cause this? Info below:

<details>
<summary>Import Settings</summary>

```
{
    "templatefile": {},
    "settings": {
        "jsonName": "id",
        "jsonNamePath": false,
        "jsonUrl": "",
        "folderName": "03_Permanent/Testing",
        "topField": "libraryVolumes",
        "notePrefix": "Journal2_",
        "noteSuffix": "",
        "handleExistingNote": 0,
        "forceArray": false,
        "multipleJSON": true,
        "uniqueNames": false
    }
}
```

</details> 
<details>
<summary>Modified Journal Template</summary>

```
---
fileClass: Journal
Journal Type: "[[120g Paper Edition]]"
{{#gt @importSourceIndex 8}}
{{#gt @importSourceIndex 98}}
BuJo Name: J{{add @importSourceIndex 1}}
{{else}}
BuJo Name: J0{{add @importSourceIndex 1}}
{{/gt}}
{{else}}
BuJo Name: J00{{add @importSourceIndex 1}}
{{/gt}}
juid: {{id}}
Start Date: {{startDate}}
End Date: {{endDate}}
{{#is firstPageSide "left"}}
First Page Left: true
{{else}}
First Page Left: false
{{/is}}
Pages: {{pageCount}}
created: {{coverPage.created}}
Cover: "[[{{coverPage.image.id}}.jpg]]"
---
{{@importSourceIndex}}
{{@importDataRoot}}
![[{{coverPage.image.id}}.jpg]]
```

</details>
<details>
<summary>Result Example:</summary>

```
---
fileClass: Journal
Journal Type: "[[120g Paper Edition]]"
BuJo Name: J001
juid: 0f891fa7-8d82-4946-b866-5ad3e72b6d00
Start Date: 2013-11
End Date: 2014-12
First Page Left: false
Pages: 240
created: 2024-04-13T18:24:39Z
Cover: "[[2c50a9b3-6941-4e5a-83f7-192389c035d8.jpg]]"
---
0
[object Object]
![[2c50a9b3-6941-4e5a-83f7-192389c035d8.jpg]]
```

</details>

**Reply by farling42** [reply, marked as answer]
There is a bug which means `@importSourceIndex` doesn't work. You can use `SourceIndex` for now, until I fix the bug.

**Reply by Truncated** [reply]
Hooray! That worked and I'm not crazy. lol
Thank you for the quick assist!

**Reply by Truncated** [reply]
Side question: Is it possible to make a calculated result the note name?
Is setvar intended to be used in that way, where I could have the resulting value of "J" or "J0" or "J00" + SourceIndex + 1 be made into a variable itself and use that for the file name?

If you already know that's not possible, I'd rather not spend the time poking at it. :)

**Reply by farling42** [reply]
Have a look at the readme which describes how to write some javascript within `${...} ` in the name field. This can be used to calculate the name from fields in the JSON field.

**Reply by Truncated** [reply]
Dangit. I can hack around with handlebars pretty quickly, but writing javascript is still a google-fest for me atm. Thank you for confirming my suspicions about that angle. At least I can cut to the biscuits when I have more time than go down a dry well.

Cheers!

**Reply by farling42** [reply]
Probably something like:
```
${ return ('000'+ (this.sourceIndex+1) ).slice(-3)  }
```

**Reply by Truncated** [reply]
Thank you for this code. I tried it as this in the name field:
```
@{${ return 'J' + ('000'+ (this.sourceIndex+1) ).slice(-3)  }}
```
and it returned Uncaught (in promise) SyntaxError: Unexpected token '{' at the 3rd character.
Neither removing the @ wrapper or the $ wrapper allowed it to function.

The only iteration that imported anything was `@{return ('000'+ (this.sourceIndex+1) ).slice(-3)}`, which gave me NaN+1, so the first is NaN and the rest are NaN1, NaN2, etc.

**Reply by Truncated** [reply]
hah! Capitals.
`@{return ('000'+ (this.SourceIndex+1) ).slice(-3)}` works. Thanks again!

**Comment by farling42** [comment]
Version 0.36.3 should have `@importSourceIndex` being set.
